tracker/internal/graphstore/v1beta/driver.go

```diff
@@ -0,0 +1,96 @@
+package v1beta
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/jmoiron/sqlx"
+
+	"gorm.io/driver/mysql"
+	"gorm.io/driver/postgres"
+	"gorm.io/driver/sqlite"
+
+	"gorm.io/gorm"
+)
+
+// Driver represents a generic interface for storing a graph.
+type Driver interface {
+	Put(ctx context.Context, data []*GraphData) error
+	Delete(ctx context.Context, data []*GraphData) error
+	List(ctx context.Context, kind string, offset, limit int) ([]*GraphData, bool, error)
+	NeighborsTo(ctx context.Context, to *GraphData) ([]*GraphData, error)
+	NeighborsFrom(ctx context.Context, from *GraphData) ([]*GraphData, error)
+}
+
+// Resolve takes in connection criteria and returns the appropriate driver
+func Resolve(driver, storageAddress, storageReadOnlyAddress string) (Driver, error) {
+	var statements *Statements
+	var dialectorRW gorm.Dialector
+	var dialectorRO gorm.Dialector
+
+	switch driver {
+	case "mysql":
+		driver = "mysql"
+		statements = MySQLStatements
+		dialectorRW = mysql.Open(storageAddress)
+		dialectorRO = mysql.Open(storageReadOnlyAddress)
+		break
+	case "sqlite", "sqlite3":
+		driver = "sqlite3"
+		statements = SQLiteStatements
+		dialectorRW = sqlite.Open(storageAddress)
+		dialectorRO = sqlite.Open(storageReadOnlyAddress)
+		break
+	case "postgres", "postgresql", "pgx":
+		driver = "postgres"
+		statements = PostgreSQLStatements
+		dialectorRW = postgres.Open(storageAddress)
+		dialectorRO = postgres.Open(storageReadOnlyAddress)
+		break
+	default:
+		return nil, fmt.Errorf("failed to resolve driver: %s", driver)
+	}
+
+	var dbrw *sql.DB
+	if len(storageAddress) > 0 {
+		gormRW, err := gorm.Open(dialectorRW, &gorm.Config{})
+		if err != nil {
+			return nil, err
+		}
+
+		// only migrate on RW connections
+		err = gormRW.AutoMigrate(&GraphData{})
+		if err != nil {
+			return nil, err
+		}
+
+		dbrw, err = gormRW.DB()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	dbro := dbrw
+	if len(storageReadOnlyAddress) > 0 {
+		gormRO, err := gorm.Open(dialectorRO, &gorm.Config{})
```

Comment:
So ORM is the way forward for Tracker to interact with DBs?

Reply:
We still use sqlx for querying but we use ORM to manage the schema migrations. I was trying to figure out if there was a way to just use GORM, but found myself spending way too much time trying to re-invent some simple sql queries. (such as life with an ORM.)